log/syslogify.1.scd
SYSLOGIFY(1)

# NAME

syslogify - transforms raw logs into syslog entries.

# SYNOPSIS

syslogify [*-h*]

syslogify [*-s*] [*-i* _ident_] [*-f* _facility_] [*-l* _level_] _regex_ _log_...

syslogify [*-s*] [*-I*] [*-i* _ident_] [*-f* _facility_] [*-l* _level_] _regex_

# OPTIONS

_regex_
	is the _regex_(3) used to parse the raw logs.
	It must have two groups.
	The first contains the log level.
	The second is the content logged.
	Everything else is ignored.
	If you use an alternate (*|*), each part must have exactly two groups, e.g.
		```
		(?:s6-rc: (\\w+): (.*))|(?:\\d{4}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2} (\\w+) (.*))
		```

_log_
	is the log to use.
	Mutually exclusive with *-I*.

*-I*
	read the log from _stdin_(3).
	Mutually exclusive with _log_.

*-s* strip escape codes from the input.

*-i* _ident_
	set the ident of the entry.

*-f* _facility_
	set the facility containing the entry (do not start with LOG_).
	This option is case insensitive.

*-l* _level_
	set a custom name for a log level.
	_level_ must follow this pattern:
		_name_:_syslog_
	where _name_ is the name used in the log and _syslog_ is the _syslog_(3) name (but do not start with LOG_).
	This option is case insensitive.

*-h* prints the help.

# EXAMPLES

```
syslogify "(\w+) (.+)" INFO This is an example
```

logs "This is an example" in the user's facility with log level INFO.

```
syslogify -i test -f news "(\w+) (.+)" INFO This is an example
```

logs "This is an example" in the news' facility with log level INFO with the ident "test".

```
syslogify -l warn:warning -l trace:debug "(\w+) (.+)" WARN This is a warning
```

logs "This is a warning" in the user's facility with log level WARNING.

```
python syslogify.py -s -l warn:warning ".* \[(\w+)] (.*)" $(printf "\\e[38;5;244m2026-04-24 14:10:29 \\e[0m[\\e[38;5;11mWARN\\e[0m] \\e[38;5;244minternal/mux.go:93 - \\e[0m\\e[38;5;11minvalid request\\e[0m\\e[38;5;244m uri=/.git/config status=429 duration=4.248µs \\e[0m")
```

logs "internal/mux.go:93 - invalid request uri=/.git/config status=429 duration=4.248µs" in the user's facility with log
level WARNING.

```
python syslogify.py '(?:s6-rc: (\\w+): (.*))|(?:\\d{4}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2} (\\w+) (.*))' s6-rc: info: service legacy-services successfully started
```

logs "service legacy-services successfully started" while

```
python syslogify.py '(?:s6-rc: (\\w+): (.*))|(?:\\d{4}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2} (\\w+) (.*))' 2026/05/22 22:02:43 INFO knot/server/db: migration applied successfully migration=add-repo-name-to-repo-keys
```

logs "knot/server/db: migration applied successfully migration=add-repo-name-to-repo-keys"

# SEE ALSO

_syslog_(3), _regex_(3)

# AUTHORS

Maintained by William Hergès <@anhgelus.world>.


log/syslogify.py
#!/usr/bin/env python3
import sys
import syslog
import argparse
import re

facilities = {
    "daemon": syslog.LOG_DAEMON,
    "user": syslog.LOG_USER,
    "auth": syslog.LOG_AUTH,
    "news": syslog.LOG_NEWS,
    "authpriv": syslog.LOG_AUTHPRIV,
    "mail": syslog.LOG_MAIL,
    "local0": syslog.LOG_LOCAL0,
    "local1": syslog.LOG_LOCAL1,
    "local2": syslog.LOG_LOCAL2,
    "local3": syslog.LOG_LOCAL3,
    "local4": syslog.LOG_LOCAL4,
    "local5": syslog.LOG_LOCAL5,
    "local6": syslog.LOG_LOCAL6,
    "local7": syslog.LOG_LOCAL7,
}

levels = {
    "debug": syslog.LOG_DEBUG,
    "info": syslog.LOG_INFO,
    "notice": syslog.LOG_NOTICE,
    "warning": syslog.LOG_WARNING,
    "err": syslog.LOG_ERR,
    "crit": syslog.LOG_CRIT,
    "alert": syslog.LOG_ALERT,
    "emerg": syslog.LOG_EMERG,
}

parser = argparse.ArgumentParser(
    "syslogify",
    description="Transforms raw log output into valid syslog entries.",
    epilog="See syslogify(1) for more information.")
parser.add_argument("-i", "--ident", help="set the ident of the syslog entry.")
parser.add_argument("-f", "--facility",
                    choices=facilities.keys(),
                    help="set the facility of the syslog entry.",
                    default="user")
parser.add_argument("-l", "--level",
                    help="set the string representation of a syslog level.",
                    action="append")
parser.add_argument("-s", "--strip", action="store_true",
                    help="strip escape codes.")
input_group = parser.add_mutually_exclusive_group(required=True)
input_group.add_argument("-I", "--stdin", action="store_true",
                         help="takes the input from stdin.")
parser.add_argument("regex", help="regex to use for parsing the input.")
input_group.add_argument(
    "log", nargs="*", help="log to convert.")
args = parser.parse_args()

escape = re.compile("\x1b"+r"\[[0-9;]*[mGKHF]")


def handle(log):
    if args.strip:
        log = escape.sub("", log)

    if len(log) == 0:
        return

    res: re.Match[str] | None = reg.match(log)
    if res == None:
        syslog.syslog(syslog.LOG_WARNING, "cannot parse: " + log)
        return

    groups = res.groups()

    if len(groups) % 2 != 0 or len(groups) == 0:
        syslog.syslog(syslog.LOG_WARNING, "invalid regex: " +
                      args.regex + ", it must have exactly 2(n+1) groups")
        exit(1)

    while len(groups) > 0 and groups[0] == None:
        groups = groups[2:]

    if len(groups) == 0:
        syslog.syslog(syslog.LOG_WARNING, "cannot extract groups from " + log)
        return

    raw_level = groups[0]
    content = groups[1]

    level = levels.get(raw_level.lower())
    if level == None:
        syslog.syslog(syslog.LOG_WARNING, "invalid log level: " + raw_level)
        exit(1)

    syslog.syslog(level, content)


for v in args.level if args.level != None else []:
    [k, val] = v.split(":", maxsplit=1)
    levels[k.lower()] = levels[val.lower()]

if args.ident != None:
    syslog.openlog(ident=args.ident,
                   facility=facilities[args.facility.lower()])
else:
    syslog.openlog(facility=facilities[args.facility.lower()])

reg: re.Pattern = re.compile(args.regex)

if args.stdin:
    for log in sys.stdin:
        handle(log)
else:
    handle(" ".join(args.log))
